torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -84,9 +90,40 @@ const ::gloo::ReductionFunction<T>* reductionFunction(const ReduceOp& r) {
   throw std::runtime_error("Unhandled ReduceOp");
 }
 
+std::vector<cudaStream_t> getStreamVector(AlgorithmEntry& entry) {
+  std::vector<cudaStream_t> streams(entry.streams.size());
+  for (auto i = 0; i < entry.streams.size(); i++) {
+    streams[i] = entry.streams[i].getStream();
+  }
+  return streams;
+}
+
+// synchronizeStreams ensures that the private streams associated with
+// an algorithm entry wait for the public streams to complete.
+void synchronizeStreams(THCState* thcState, AlgorithmEntry* entry) {
+  CUDADevice deviceGuard;
+  const auto& key = entry->key;
+  for (auto i = 0; i < key.devices.size(); i++) {
+    const auto& device = key.devices[i];
```

Comment:
maybe you can also do
```
deviceGuard.setDevice(device);
auto publicStream = THCState_getCurrentStream(thcState):
```

Reply:
I prefer the current version since it does 1 less cudaGetDevice call (even though it is cheap)